### PR TITLE
`r\iothub`: Support managed service identity

### LINKED ISSUE #9139: Support for IoTHub Managed Service Identity for Endpoints
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The IoTHub supports the Managed Service Identity Feature to use the MSI for Authorization against endpoints.
* Currently it is not possible to create an IoT Hub with identity block.
* Currently it is not possible to use the Identity for Authorization when creating IoT Hub Endpoints

### New or Affected Resource(s)

* azurerm_iothub

### Potential Terraform Configuration

```hcl
resource "azurerm_iothub" "test" {
# [...]
  identity {
     type = "SystemAssigned"
  }
  
  endpoint {
    type = "AzureIoTHub.ServiceBusTopic"
    # [...]
    use_managed_identity = true
  }
```
This should create a new Role Assignment for the MSI. Alternative it could be possible to create an azurerm_iothub_endpoint resource.

### References

No References found

### PR DESCRIPTION
**Pull request by myc2h6o**
Fix #9139
This is required when Iot Hub needs to talk to other resources on a private v-net. Support for `AuthenticatinType` in routing, file upload, device import/export will be added in following prs
Test result:
image

### CODE REVIEW

File: internal/services/iothub/iothub_resource.go [hunk 1]
```diff
@@ -1288,6 +1309,59 @@ func flattenIPFilterRules(in *[]devices.IPFilterRule) []interface{} {
...
+		return nil, err
+	}
+
+	var identityIds map[string]*devices.ArmUserIdentity
+	if len(config.UserAssignedIdentityIds) != 0 {
+		identityIds = map[string]*devices.ArmUserIdentity{}
+		for _, id := range config.UserAssignedIdentityIds {
+			identityIds[id] = &devices.ArmUserIdentity{}
+		}
+	}
```
Comment (tombuildsstuff) [inline internal/services/iothub/iothub_resource.go:1324]: this'll ensure these are always removed:

```suggestion
	identityIds = make(map[string]*devices.ArmUserIdentity, 0)
	for _, id := range config.UserAssignedIdentityIds {
		identityIds[id] = &devices.ArmUserIdentity{}
	}
```

File: website/docs/r/iothub.html.markdown [hunk 2]
```diff
@@ -188,6 +190,16 @@ An `endpoint` block supports the following:
 
 ---
 
+A `identity` block supports the following:
+
+* `type` - (Required) The type of Managed Identity which should be assigned to the Iot Hub. Possible values are `SystemAssigned`, `UserAssigned` and `SystemAssigned, UserAssigned`.
+
+* `identity_ids` - (Optional) A list of User Managed Identity ID's which should be assigned to the Iot Hub.
+
+~> **NOTE:** This is required when `type` is set to `UserAssigned`.
```
Comment (tombuildsstuff) [inline website/docs/r/iothub.html.markdown:199]: this is also required when it's set to SystemAssigned/UserAssigned:

```suggestion
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hi @myc2h6o 

Thanks for this PR.

I've taken a look through and left some comments inline, but if we can fix those up then we can take another look.

Thanks!

File: website/docs/r/iothub.html.markdown (hunk 2 above)
Comment (myc2h6o) [reply website/docs/r/iothub.html.markdown]: Updated

File: internal/services/iothub/iothub_resource.go (hunk 1 above)
Comment (myc2h6o) [reply internal/services/iothub/iothub_resource.go]: Have updated the code, but I didn't fully understand the comment, is it related to the garbage collection mechanism?

And I found that identityIds has to be nil for a SystemAssigned identity. An error will be thrown if an empty map is given. So I've updated the code to first create the `ArmUserIdentity` without `identityIds` and set it only when provided. Please check if it looks good.

**Comment by myc2h6o** [comment]
@tombuildsstuff thanks for reviewing the change. I've updated the code, please take another look.

**Comment by myc2h6o** [comment]
Rebased to latest main to resolve conflict

**Comment by myc2h6o** [comment]
Hi @tombuildsstuff I saw you added new identity types in the linked pr, I've updated this change to use the new  `commonschema.SystemAssignedUserAssignedIdentity()` similar to #14007, please check if this looks good.

**Review by katbyte — APPROVED** [review]
Thanks @myc2h6o - LGTM 💎

**Review by tombuildsstuff — APPROVED** [review]
LGTM - thanks for pushing those changes @myc2h6o

**Comment by tombuildsstuff** [comment]
Tests pass:

Screenshot 2021-12-20 at 11 22 06

**Pull request merged**